src/dir/glob.cr

```diff
@@ -1,266 +1,358 @@
+  #
+  # NOTE: Path separator in patterns is `/` even on Windows. The returned file names use system-specific path separators.
   def self.[](*patterns) : Array(String)
     glob(patterns)
   end
 
+  # ditto
   def self.[](patterns : Enumerable(String)) : Array(String)
     glob(patterns)
   end
 
-  def self.glob(*patterns) : Array(String)
-    glob(patterns)
-  end
-
-  def self.glob(*patterns)
-    glob(patterns) do |pattern|
-      yield pattern
-    end
+  # Returns an array of all files that match against any of *patterns*.
+  #
+  # The pattern syntax is similar to shell filename globbing. It may contain the following metacharacters:
+  #
+  # * `*` matches an unlimited number of arbitrary characters.
+  #   * `"*"` matches all regular files.
+  #   * `"c*"` matches all files beginning with `c`.
+  #   * `"*c"` matches all files ending with `c`.
+  #   * `"*c*"` matches all files that have `c` in them (including at the beginning or end).
+  # * `**` matches directories recursively or files expansively.
+  # * `?` matches any one character.
+  # * `{a,b}` matches subpattern `a` or `b`.
+  #
+  # **options:**
+  # * *allow_dots*: Match hidden files if `true` (default: `false`).
+  #
+  # NOTE: Path separator in patterns is `/` even on Windows. The returned file names use system-specific path separators.
+  def self.glob(*patterns, **options) : Array(String)
+    glob(patterns, **options)
   end
 
-  def self.glob(patterns : Enumerable(String)) : Array(String)
+  # ditto
+  def self.glob(patterns : Enumerable(String), **options) : Array(String)
     paths = [] of String
-    glob(patterns) do |path|
+    glob(patterns, **options) do |path|
       paths << path
     end
     paths
   end
 
-  def self.glob(patterns : Enumerable(String))
-    special = {'*', '?', '{', '}'}
-    cwd = self.current
-    root = File::SEPARATOR_STRING
-    patterns.each do |ptrn|
-      next if ptrn.empty?
-
-      starts_with_path_separator = ptrn.starts_with?(File::SEPARATOR)
-
-      # If the pattern ends with a file separator we only want leaf
-      # directories: regex is the same as the one without that last char
-      wants_dir = false
-      if ptrn.ends_with?(File::SEPARATOR)
-        ptrn = ptrn.rchop
-        wants_dir = true
-      end
+  # Yields all files that match against any of *patterns*.
+  #
+  # The pattern syntax is similar to shell filename globbing. It may contain the following metacharacters:
+  #
+  # * `*` matches an unlimited number of arbitrary characters.
+  #   * `"*"` matches all regular files.
+  #   * `"c*"` matches all files beginning with `c`.
+  #   * `"*c"` matches all files ending with `c`.
+  #   * `"*c*"` matches all files that have `c` in them (including at the beginning or end).
+  # * `**` matches directories recursively or files expansively.
+  # * `?` matches any one character.
+  # * `{a,b}` matches subpattern `a` or `b`.
+  #
+  # **options:**
+  # * *allow_dots*: Match hidden files if `true` (default: `false`).
```

Comment:
Let's keep the `options`. I was thinking of maybe using an enum but options is nicer here, simpler to specify. I'm not sure about the name, though. In C it's called `FNM_PERIOD`, in Ruby it's `FNM_DOTMATCH` but what it means is that it matches hidden files, or files that start with a dot/period. `allow_dots` suggests either the path or pattern allows dots anywhere. Using "hidden files" isn't totally correct either because in Windows that's a different thing. Maybe just `match_leading_dot` or similar?

Reply:
Sounds good :+1:

Reply:
Just a reminder that `allow_dots` still needs to be renamed to `match_leading_lot`.

In fact, since before this PR we didn't have this option, you could actually remove it for now. We can always add it later. I don't know why `.foo` files should be treated differently.

Reply:
Yeah, sure. I was just waiting if maybe someone would come up with a better name.
What do you think about `recurse_hidden_files`? I think this would be more descriptive.

Reply:
The current implementation matches hidden files, but the default should be *not* to. So that's a breaking change. We should have this option right now to allow people to keep the existing behaviour.

Reply:
Why not? I just tried this in Go:

```go
package main

import "path/filepath"
import "fmt"

func main() {
	matches, _ := filepath.Glob("./*")
	fmt.Println(matches)
}
```

Here's part of the output:

```
[.build .dockerignore .editorconfig .git .gitignore .travis.yml ...
```

Go doesn't hide dot files by default. Ruby does. That doesn't mean either is correct. But I don't understand why dot files are so important (or not so important) that they should be hidden.

I think we should start with no options, showing all files. Later, in a separate issue/PR, we can discuss options to add to this method.